File: paradas-programadas/js/state.js
/*
 * Estado da aplicação em memória + operações de CRUD.
 * Toda alteração passa por `persist()`, que salva no IndexedDB / pasta
 * escolhida e dispara `onChange` para a UI se atualizar.
 */

const State = (() => {
  let data = criarWorkspaceVazio();
  let listeners = [];

  function criarWorkspaceVazio() {
    const cal = calendarioPadrao();
    return {
      version: 1,
      paradaAtivaId: null,
      paradas: [],
      calendarios: [cal],
      atividades: []
    };
  }

  function onChange(fn) { listeners.push(fn); }
  function notify() { listeners.forEach(fn => fn()); }

  function persist() {
    Storage.save(data);
    notify();
  }

  async function init() {
    await Storage.init();
    const loaded = await Storage.load();
    if (loaded && loaded.paradas) {
      data = migrar(loaded);
    }
    await Storage.restaurarPastaSalva();
    notify();
  }

  function migrar(d) {
    if (!d.calendarios || d.calendarios.length === 0) d.calendarios = [calendarioPadrao()];
    if (!d.atividades) d.atividades = [];
    if (!d.paradas) d.paradas = [];
    d.atividades.forEach(a => {
      if (!Array.isArray(a.imagens)) a.imagens = [];
      if (!Array.isArray(a.recursos)) a.recursos = [];
      if (a.progresso === undefined) a.progresso = 0;
      if (!a.status) a.status = 'planejada';
      if (a.inicioReal === undefined) a.inicioReal = null;
      if (a.fimReal === undefined) a.fimReal = null;
      if (a.predecessoraId === undefined) a.predecessoraId = null;
      if (a.defasagemHoras === undefined) a.defasagemHoras = 0;
    });
    return d;
  }

  function substituirTudo(novoData) {
    data = migrar(novoData);
    persist();
  }

  function getData() { return data; }

  // ---------- Paradas ----------

  function listarParadas() { return data.paradas.slice().sort((a, b) => (a.criadoEm || '').localeCompare(b.criadoEm || '')); }

  function getParada(id) { return data.paradas.find(p => p.id === id) || null; }

  function getParadaAtiva() { return getParada(data.paradaAtivaId); }

  function setParadaAtiva(id) {
    data.paradaAtivaId = id;
    persist();
  }

  function salvarParada(parada) {
    if (!parada.calendarioId && data.calendarios[0]) parada.calendarioId = data.calendarios[0].id;
    const idx = data.paradas.findIndex(p => p.id === parada.id);
    if (idx >= 0) {
      data.paradas[idx] = parada;
    } else {
      parada.id = parada.id || uid();
      parada.criadoEm = new Date().toISOString();
      data.paradas.push(parada);
      if (!data.paradaAtivaId) data.paradaAtivaId = parada.id;
    }
    recalcularProgramacao(parada.id);
    persist();
    return parada;
  }

  function excluirParada(id) {
    data.paradas = data.paradas.filter(p => p.id !== id);
    data.atividades = data.atividades.filter(a => a.paradaId !== id);
    if (data.paradaAtivaId === id) {
      data.paradaAtivaId = data.paradas[0] ? data.paradas[0].id : null;
    }
    persist();
  }

  // ---------- Calendários ----------

  function listarCalendarios() { return data.calendarios; }
  function getCalendario(id) { return data.calendarios.find(c => c.id === id) || data.calendarios[0]; }

  function salvarCalendario(cal) {
    const idx = data.calendarios.findIndex(c => c.id === cal.id);
    if (idx >= 0) data.calendarios[idx] = cal;
    else { cal.id = cal.id || uid(); data.calendarios.push(cal); }
    // qualquer parada usando esse calendário precisa recalcular datas
    data.paradas.filter(p => p.calendarioId === cal.id).forEach(p => recalcularProgramacao(p.id));
    persist();
    return cal;
  }

  function excluirCalendario(id) {
    if (data.calendarios.length <= 1) {
      showToast('É preciso manter ao menos um calendário.', true);
      return false;
    }
    const emUso = data.paradas.some(p => p.calendarioId === id);
    if (emUso) {
      showToast('Este calendário está em uso por uma parada e não pode ser excluído.', true);
      return false;
    }
    data.calendarios = data.calendarios.filter(c => c.id !== id);
    persist();
    return true;
  }

  // ---------- Atividades ----------

  /** Ordenada por Data/Hora de Início (sem data vai para o final); empate usa a ordem de cadastro. */
  function listarAtividadesDaParada(paradaId) {
    return data.atividades.filter(a => a.paradaId === paradaId).sort((a, b) => {
      const da = a.dataInicio ? new Date(a.dataInicio).getTime() : Infinity;
      const db = b.dataInicio ? new Date(b.dataInicio).getTime() : Infinity;
      if (da !== db) return da - db;
      return (a.ordem || 0) - (b.ordem || 0);
    });
  }

  /** Retorna apenas as atividades de topo (sem parentId), cada uma com .subAtividades preenchido.
   *  Tanto as atividades de topo quanto as sub-atividades de cada uma ficam em ordem
   *  cronológica (Data/Hora de Início), e cada sub-atividade permanece agrupada
   *  logo abaixo da sua atividade correspondente. */
  function arvoreAtividades(paradaId) {
    const todas = listarAtividadesDaParada(paradaId);
    const porId = {};
    todas.forEach(a => { porId[a.id] = { ...a, subAtividades: [] }; });
    const raizes = [];
    todas.forEach(a => {
      if (a.parentId && porId[a.parentId]) {
        porId[a.parentId].subAtividades.push(porId[a.id]);
      } else {
        raizes.push(porId[a.id]);
      }
    });
    return raizes;
  }

  /** Achata a árvore em uma lista [{ ...atividade, nivel }], mantendo cada
   *  sub-atividade logo após a atividade correspondente (ordem de exibição
   *  usada pela listagem, pelo Gantt e pelos relatórios). */
  function listaAchatada(paradaId) {
    const out = [];
    (function visitar(lista, nivel) {
      lista.forEach(a => {
        const { subAtividades, ...resto } = a;
        out.push({ ...resto, nivel });
        visitar(subAtividades || [], nivel + 1);
      });
    })(arvoreAtividades(paradaId), 0);
    return out;
  }

  function getAtividade(id) { return data.atividades.find(a => a.id === id) || null; }

  function calendarioDaAtividade(atividade) {
    const parada = getParada(atividade.paradaId);
    return parada ? getCalendario(parada.calendarioId) : calendarioPadrao();
  }

  /** Duração real (horas produtivas), ou null se início e/ou fim real não estiverem preenchidos. */
  function duracaoRealHoras(atividade) {
    if (!atividade.inicioReal || !atividade.fimReal) return null;
    const calendario = calendarioDaAtividade(atividade);
    return calcularDuracaoHoras(new Date(atividade.inicioReal), new Date(atividade.fimReal), calendario);
  }

  /**
   * Salva uma atividade. `modoCalculo` indica qual campo foi editado por
   * último pelo usuário: 'duracao' (recalcula dataFim) ou 'fim' (recalcula
   * duracaoHoras). Default: 'duracao'.
   */
  function salvarAtividade(atividade, modoCalculo = 'duracao') {
    // Com predecessora definida, a Data Início é derivada automaticamente pela
    // cadeia (recalcularProgramacao, logo abaixo) — o que foi digitado aqui é ignorado.
    if (!atividade.predecessoraId) {
      const calendario = calendarioDaAtividade(atividade);
      const inicio = new Date(atividade.dataInicio);
      if (modoCalculo === 'fim' && atividade.dataFim) {
        atividade.duracaoHoras = calcularDuracaoHoras(inicio, new Date(atividade.dataFim), calendario);
      } else {
        const fim = calcularDataFim(inicio, Number(atividade.duracaoHoras) || 0, calendario);
        atividade.dataFim = fim ? fim.toISOString() : null;
      }
    }

    const idx = data.atividades.findIndex(a => a.id === atividade.id);
    if (idx >= 0) {
      data.atividades[idx] = atividade;
    } else {
      atividade.id = atividade.id || uid();
      atividade.ordem = atividade.ordem ?? (listarAtividadesDaParada(atividade.paradaId).length + 1);
      if (!Array.isArray(atividade.imagens)) atividade.imagens = [];
      if (!Array.isArray(atividade.recursos)) atividade.recursos = [];
      data.atividades.push(atividade);
    }
    recalcularProgramacao(atividade.paradaId);
    persist();
    return atividade;
  }

  function excluirAtividade(id) {
    const alvo = getAtividade(id);
    // exclui também as sub-atividades
    const idsParaExcluir = new Set([id]);
    let mudou = true;
    while (mudou) {
      mudou = false;
      data.atividades.forEach(a => {
        if (a.parentId && idsParaExcluir.has(a.parentId) && !idsParaExcluir.has(a.id)) {
          idsParaExcluir.add(a.id);
          mudou = true;
        }
      });
    }
    // atividades que dependiam de algo removido passam a ter data/hora manual (mantém o último valor calculado)
    data.atividades.forEach(a => {
      if (a.predecessoraId && idsParaExcluir.has(a.predecessoraId)) a.predecessoraId = null;
    });
    data.atividades = data.atividades.filter(a => !idsParaExcluir.has(a.id));
    if (alvo) recalcularProgramacao(alvo.paradaId);
    persist();
  }

  // ---------- Predecessoras / sucessoras ----------

  /** Atividades (da mesma parada) que têm `atividadeId` como predecessora direta. */
  function sucessorasDiretas(atividadeId) {
    return data.atividades.filter(a => a.predecessoraId === atividadeId);
  }

  /** IDs de `atividadeId` + todas as suas sucessoras diretas/indiretas (cadeia completa). */
  function cadeiaSucessoras(atividadeId, visitados = new Set()) {
    if (visitados.has(atividadeId)) return visitados;
    visitados.add(atividadeId);
    sucessorasDiretas(atividadeId).forEach(s => cadeiaSucessoras(s.id, visitados));
    return visitados;
  }

  /**
   * Recalcula Data Início/Fim de todas as atividades de uma parada, respeitando a
   * cadeia de predecessoras: quem tem predecessora tem sua Data Início derivada
   * automaticamente de "Data Fim da predecessora + defasagem"; quem não tem
   * predecessora mantém a Data Início que foi digitada manualmente. Percorre a
   * cadeia em profundidade (predecessora antes de sucessora) para propagar
   * mudanças em cascata, e se protege contra referência circular.
   */
  function recalcularProgramacao(paradaId) {
    const parada = getParada(paradaId);
    const calendario = parada ? getCalendario(parada.calendarioId) : calendarioPadrao();
    const todas = listarAtividadesDaParada(paradaId);
    const porId = new Map(todas.map(a => [a.id, a]));
    const processadas = new Set();

    function processar(a, pilha) {
      if (!a || processadas.has(a.id) || pilha.has(a.id)) return;
      pilha.add(a.id);
      if (a.predecessoraId && porId.has(a.predecessoraId)) {
        const pred = porId.get(a.predecessoraId);
        processar(pred, pilha);
        if (pred.dataFim) {
          const lag = Number(a.defasagemHoras) || 0;
          a.dataInicio = new Date(new Date(pred.dataFim).getTime() + lag * 3600000).toISOString();
        }
      }
      const fim = a.dataInicio ? calcularDataFim(new Date(a.dataInicio), Number(a.duracaoHoras) || 0, calendario) : null;
      a.dataFim = fim ? fim.toISOString() : null;
      processadas.add(a.id);
      pilha.delete(a.id);
    }

    todas.forEach(a => processar(a, new Set()));
  }

  // ---------- Estatísticas ----------

  function faixaDataParada(paradaId) {
    const ativs = listarAtividadesDaParada(paradaId);
    if (ativs.length === 0) return { inicio: null, fim: null };
    let inicio = null, fim = null;
    ativs.forEach(a => {
      if (a.dataInicio) {
        const d = new Date(a.dataInicio);
        if (!inicio || d < inicio) inicio = d;
      }
      if (a.dataFim) {
        const d = new Date(a.dataFim);
        if (!fim || d > fim) fim = d;
      }
    });
    return { inicio, fim };
  }

  return {
    init, onChange, persist, getData, substituirTudo,
    listarParadas, getParada, getParadaAtiva, setParadaAtiva, salvarParada, excluirParada,
    listarCalendarios, getCalendario, salvarCalendario, excluirCalendario,
    listarAtividadesDaParada, arvoreAtividades, listaAchatada, getAtividade, salvarAtividade, excluirAtividade,
    calendarioDaAtividade, faixaDataParada, duracaoRealHoras,
    sucessorasDiretas, cadeiaSucessoras, recalcularProgramacao
  };
})();

File: paradas-programadas-python/app/state.py
"""
Estado da aplicação em memória + operações de CRUD.
Port fiel de state.js (mesma lógica, mesmos nomes de campo no JSON — o que
torna o `dados-paradas.json` desta versão em Python intercambiável com o
da versão web do mesmo sistema).

`AppState.data` é sempre um dict puro, pronto para `json.dump()` sem
nenhuma etapa extra de serialização: datas são guardadas como strings ISO
8601 em UTC (ver `dates.py`), exatamente como a versão web grava.
"""

from __future__ import annotations
import uuid
from datetime import datetime, timedelta

from . import dates
from .calendar_engine import calcular_data_fim, calcular_duracao_horas, calendario_padrao

STATUS_LABELS = {
    'planejada': 'Planejada',
    'em_andamento': 'Em andamento',
    'concluida': 'Concluída',
    'atrasada': 'Atrasada',
}

STATUS_CORES = {
    'planejada': '#4338ca',
    'em_andamento': '#b45309',
    'concluida': '#15803d',
    'atrasada': '#b91c1c',
}

STATUS_CORES_CLARAS = {
    'planejada': '#e0e7ff',
    'em_andamento': '#fef3c7',
    'concluida': '#dcfce7',
    'atrasada': '#fee2e2',
}

TIPOS_RECURSO = {
    'mao_de_obra': 'Mão de obra',
    'equipamento': 'Equipamento',
    'material': 'Material',
    'servico': 'Serviço',
}


def novo_id() -> str:
    return str(uuid.uuid4())


class AppState:
    def __init__(self):
        self.data: dict = self._workspace_vazio()
        self._listeners = []
        self.on_erro = None  # callback opcional: on_erro(mensagem)
        self.on_persist = None  # callback opcional: on_persist(data) -> grava em disco

    # ---------- infraestrutura ----------

    def _workspace_vazio(self) -> dict:
        return {
            'version': 1,
            'paradaAtivaId': None,
            'paradas': [],
            'calendarios': [calendario_padrao()],
            'atividades': [],
        }

    def on_change(self, fn):
        self._listeners.append(fn)

    def _notify(self):
        for fn in list(self._listeners):
            fn()

    def _avisar(self, mensagem: str):
        if self.on_erro:
            self.on_erro(mensagem)

    def carregar(self, novo_data: dict | None):
        if novo_data and novo_data.get('paradas') is not None:
            self.data = self._migrar(novo_data)
        self._notify()

    def substituir_tudo(self, novo_data: dict):
        self.data = self._migrar(novo_data)
        self._notify()

    def _migrar(self, d: dict) -> dict:
        if not d.get('calendarios'):
            d['calendarios'] = [calendario_padrao()]
        d.setdefault('atividades', [])
        d.setdefault('paradas', [])
        d.setdefault('paradaAtivaId', None)
        d.setdefault('version', 1)
        for a in d['atividades']:
            a.setdefault('imagens', [])
            a.setdefault('recursos', [])
            a.setdefault('progresso', 0)
            a.setdefault('status', 'planejada')
            a.setdefault('inicioReal', None)
            a.setdefault('fimReal', None)
            a.setdefault('predecessoraId', None)
            a.setdefault('defasagemHoras', 0)
            a.setdefault('parentId', None)
            a.setdefault('numeroOS', '')
        return d

    def get_data(self) -> dict:
        return self.data

    # ---------- Paradas ----------

    def listar_paradas(self) -> list[dict]:
        return sorted(self.data['paradas'], key=lambda p: p.get('criadoEm') or '')

    def get_parada(self, id_: str | None) -> dict | None:
        if not id_:
            return None
        return next((p for p in self.data['paradas'] if p['id'] == id_), None)

    def get_parada_ativa(self) -> dict | None:
        return self.get_parada(self.data.get('paradaAtivaId'))

    def set_parada_ativa(self, id_: str):
        self.data['paradaAtivaId'] = id_
        self._persist()

    def salvar_parada(self, parada: dict) -> dict:
        if not parada.get('calendarioId') and self.data['calendarios']:
            parada['calendarioId'] = self.data['calendarios'][0]['id']
        idx = next((i for i, p in enumerate(self.data['paradas']) if p['id'] == parada.get('id')), -1)
        if idx >= 0:
            self.data['paradas'][idx] = parada
        else:
            parada['id'] = parada.get('id') or novo_id()
            parada['criadoEm'] = dates.agora_iso_utc()
            self.data['paradas'].append(parada)
            if not self.data.get('paradaAtivaId'):
                self.data['paradaAtivaId'] = parada['id']
        self.recalcular_programacao(parada['id'])
        self._persist()
        return parada

    def excluir_parada(self, id_: str):
        self.data['paradas'] = [p for p in self.data['paradas'] if p['id'] != id_]
        self.data['atividades'] = [a for a in self.data['atividades'] if a['paradaId'] != id_]
        if self.data.get('paradaAtivaId') == id_:
            self.data['paradaAtivaId'] = self.data['paradas'][0]['id'] if self.data['paradas'] else None
        self._persist()

    # ---------- Calendários ----------

    def listar_calendarios(self) -> list[dict]:
        return self.data['calendarios']

    def get_calendario(self, id_: str | None) -> dict:
        cal = next((c for c in self.data['calendarios'] if c['id'] == id_), None)
        return cal or self.data['calendarios'][0]

    def salvar_calendario(self, cal: dict) -> dict:
        idx = next((i for i, c in enumerate(self.data['calendarios']) if c['id'] == cal.get('id')), -1)
        if idx >= 0:
            self.data['calendarios'][idx] = cal
        else:
            cal['id'] = cal.get('id') or novo_id()
            self.data['calendarios'].append(cal)
        for p in self.data['paradas']:
            if p.get('calendarioId') == cal['id']:
                self.recalcular_programacao(p['id'])
        self._persist()
        return cal

    def excluir_calendario(self, id_: str) -> bool:
        if len(self.data['calendarios']) <= 1:
            self._avisar('É preciso manter ao menos um calendário.')
            return False
        em_uso = any(p.get('calendarioId') == id_ for p in self.data['paradas'])
        if em_uso:
            self._avisar('Este calendário está em uso por uma parada e não pode ser excluído.')
            return False
        self.data['calendarios'] = [c for c in self.data['calendarios'] if c['id'] != id_]
        self._persist()
        return True

    # ---------- Atividades ----------

    def listar_atividades_da_parada(self, parada_id: str) -> list[dict]:
        """Ordenada por Data/Hora de Início (sem data vai para o final);
        empate usa a ordem de cadastro."""
        atividades = [a for a in self.data['atividades'] if a['paradaId'] == parada_id]

        def chave(a):
            di = dates.from_iso_utc(a.get('dataInicio'))
            return (di is None, di or datetime.max, a.get('ordem') or 0)

        return sorted(atividades, key=chave)

    def arvore_atividades(self, parada_id: str) -> list[dict]:
        """Atividades de topo (sem parentId), cada uma com 'subAtividades'
        preenchido recursivamente — sem limite de profundidade."""
        todas = self.listar_atividades_da_parada(parada_id)
        por_id = {a['id']: {**a, 'subAtividades': []} for a in todas}
        raizes = []
        for a in todas:
            pai_id = a.get('parentId')
            if pai_id and pai_id in por_id:
                por_id[pai_id]['subAtividades'].append(por_id[a['id']])
            else:
                raizes.append(por_id[a['id']])
        return raizes

    def lista_achatada(self, parada_id: str) -> list[dict]:
        """Achata a árvore em uma lista [{...atividade, 'nivel': n}],
        mantendo cada sub-atividade logo após a atividade correspondente."""
        out = []

        def visitar(lista, nivel):
            for a in lista:
                resto = {k: v for k, v in a.items() if k != 'subAtividades'}
                resto['nivel'] = nivel
                out.append(resto)
                visitar(a.get('subAtividades') or [], nivel + 1)

        visitar(self.arvore_atividades(parada_id), 0)
        return out

    def get_atividade(self, id_: str | None) -> dict | None:
        if not id_:
            return None
        return next((a for a in self.data['atividades'] if a['id'] == id_), None)

    def calendario_da_atividade(self, atividade: dict) -> dict:
        parada = self.get_parada(atividade.get('paradaId'))
        return self.get_calendario(parada.get('calendarioId')) if parada else calendario_padrao()

    def duracao_real_horas(self, atividade: dict) -> float | None:
        if not atividade.get('inicioReal') or not atividade.get('fimReal'):
            return None
        cal = self.calendario_da_atividade(atividade)
        ini = dates.from_iso_utc(atividade['inicioReal'])
        fim = dates.from_iso_utc(atividade['fimReal'])
        return calcular_duracao_horas(ini, fim, cal)

    def salvar_atividade(self, atividade: dict, modo_calculo: str = 'duracao') -> dict:
        """Salva uma atividade. `modo_calculo` indica qual campo foi editado
        por último: 'duracao' (recalcula dataFim) ou 'fim' (recalcula
        duracaoHoras). Ignorado quando a atividade tem predecessora — nesse
        caso a Data Início é sempre derivada pela cadeia, em
        `recalcular_programacao` logo abaixo."""
        if not atividade.get('predecessoraId'):
            cal = self.calendario_da_atividade(atividade)
            inicio = dates.from_iso_utc(atividade.get('dataInicio'))
            if modo_calculo == 'fim' and atividade.get('dataFim'):
                fim = dates.from_iso_utc(atividade['dataFim'])
                atividade['duracaoHoras'] = calcular_duracao_horas(inicio, fim, cal)
            else:
                fim = calcular_data_fim(inicio, float(atividade.get('duracaoHoras') or 0), cal)
                atividade['dataFim'] = dates.to_iso_utc(fim)

        idx = next((i for i, a in enumerate(self.data['atividades']) if a['id'] == atividade.get('id')), -1)
        if idx >= 0:
            self.data['atividades'][idx] = atividade
        else:
            atividade['id'] = atividade.get('id') or novo_id()
            if atividade.get('ordem') is None:
                atividade['ordem'] = len(self.listar_atividades_da_parada(atividade['paradaId'])) + 1
            atividade.setdefault('imagens', [])
            atividade.setdefault('recursos', [])
            self.data['atividades'].append(atividade)

        self.recalcular_programacao(atividade['paradaId'])
        self._persist()
        return atividade

    def excluir_atividade(self, id_: str):
        alvo = self.get_atividade(id_)
        # exclui também as sub-atividades, em qualquer profundidade
        ids_excluir = {id_}
        mudou = True
        while mudou:
            mudou = False
            for a in self.data['atividades']:
                if a.get('parentId') in ids_excluir and a['id'] not in ids_excluir:
                    ids_excluir.add(a['id'])
                    mudou = True
        # atividades que dependiam de algo removido voltam a ter data/hora manual
        for a in self.data['atividades']:
            if a.get('predecessoraId') in ids_excluir:
                a['predecessoraId'] = None
        self.data['atividades'] = [a for a in self.data['atividades'] if a['id'] not in ids_excluir]
        if alvo:
            self.recalcular_programacao(alvo['paradaId'])
        self._persist()

    # ---------- Predecessoras / sucessoras ----------

    def sucessoras_diretas(self, atividade_id: str) -> list[dict]:
        return [a for a in self.data['atividades'] if a.get('predecessoraId') == atividade_id]

    def cadeia_sucessoras(self, atividade_id: str, visitados: set | None = None) -> set:
        """IDs de `atividade_id` + todas as suas sucessoras diretas/indiretas."""
        if visitados is None:
            visitados = set()
        if atividade_id in visitados:
            return visitados
        visitados.add(atividade_id)
        for s in self.sucessoras_diretas(atividade_id):
            self.cadeia_sucessoras(s['id'], visitados)
        return visitados

    # ---------- Hierarquia (sub-atividades em qualquer profundidade) ----------

    def filhos_diretos(self, atividade_id: str) -> list[dict]:
        return [a for a in self.data['atividades'] if a.get('parentId') == atividade_id]

    def cadeia_descendentes(self, atividade_id: str, visitados: set | None = None) -> set:
        """IDs de `atividade_id` + todos os seus descendentes. Usado para
        impedir que uma atividade vire sub-atividade de algo que já é seu
        próprio descendente (evita laço na árvore)."""
        if visitados is None:
            visitados = set()
        if atividade_id in visitados:
            return visitados
        visitados.add(atividade_id)
        for f in self.filhos_diretos(atividade_id):
            self.cadeia_descendentes(f['id'], visitados)
        return visitados

    def recalcular_programacao(self, parada_id: str):
        """Recalcula Data Início/Fim de todas as atividades de uma parada,
        respeitando a cadeia de predecessoras: quem tem predecessora tem sua
        Data Início derivada automaticamente de "Data Fim da predecessora +
        defasagem"; quem não tem predecessora mantém a Data Início digitada
        manualmente. Percorre a cadeia em profundidade (predecessora antes
        de sucessora) para propagar mudanças em cascata, protegendo contra
        referência circular."""
        parada = self.get_parada(parada_id)
        calendario = self.get_calendario(parada.get('calendarioId')) if parada else calendario_padrao()
        todas = self.listar_atividades_da_parada(parada_id)
        por_id = {a['id']: a for a in todas}
        processadas = set()

        def processar(a: dict, pilha: set):
            if a is None or a['id'] in processadas or a['id'] in pilha:
                return
            pilha = pilha | {a['id']}
            pred_id = a.get('predecessoraId')
            if pred_id and pred_id in por_id:
                pred = por_id[pred_id]
                processar(pred, pilha)
                if pred.get('dataFim'):
                    lag = float(a.get('defasagemHoras') or 0)
                    pred_fim = dates.from_iso_utc(pred['dataFim'])
                    novo_inicio = pred_fim + timedelta(hours=lag)
                    a['dataInicio'] = dates.to_iso_utc(novo_inicio)

            inicio = dates.from_iso_utc(a.get('dataInicio'))
            fim = calcular_data_fim(inicio, float(a.get('duracaoHoras') or 0), calendario) if inicio else None
            a['dataFim'] = dates.to_iso_utc(fim)
            processadas.add(a['id'])

        for a in todas:
            processar(a, set())

    # ---------- Estatísticas ----------

    def faixa_data_parada(self, parada_id: str) -> tuple[datetime | None, datetime | None]:
        atividades = self.listar_atividades_da_parada(parada_id)
        inicio = fim = None
        for a in atividades:
            di = dates.from_iso_utc(a.get('dataInicio'))
            df = dates.from_iso_utc(a.get('dataFim'))
            if di and (inicio is None or di < inicio):
                inicio = di
            if df and (fim is None or df > fim):
                fim = df
        return inicio, fim

    def calcular_estatisticas(self, parada_id: str) -> dict:
        atividades = self.listar_atividades_da_parada(parada_id)
        total = len(atividades)
        por_status = {k: 0 for k in STATUS_LABELS}
        progresso_soma = 0.0
        horas_totais = 0.0
        recursos_por_tipo: dict[str, float] = {}
        custo_total = 0.0

        for a in atividades:
            por_status[a.get('status', 'planejada')] = por_status.get(a.get('status', 'planejada'), 0) + 1
            progresso_soma += float(a.get('progresso') or 0)
            horas_totais += float(a.get('duracaoHoras') or 0)
            for r in a.get('recursos') or []:
                tipo = r.get('tipo') or 'outro'
                recursos_por_tipo[tipo] = recursos_por_tipo.get(tipo, 0) + float(r.get('quantidade') or 0)
                custo_total += float(r.get('quantidade') or 0) * float(r.get('custoUnitario') or 0)

        progresso_medio = round(progresso_soma / total, 1) if total else 0
        inicio, fim = self.faixa_data_parada(parada_id)
        duracao_total_horas = round((fim - inicio).total_seconds() / 3600, 1) if inicio and fim else None

        return {
            'total': total,
            'porStatus': por_status,
            'progressoMedio': progresso_medio,
            'horasTotais': horas_totais,
            'recursosPorTipo': recursos_por_tipo,
            'custoTotal': custo_total,
            'inicio': inicio,
            'fim': fim,
            'duracaoTotalHoras': duracao_total_horas,
        }

    # ---------- Persistência ----------

    def _persist(self):
        if self.on_persist:
            self.on_persist(self.data)
        self._notify()
